### PR TITLE
Cpp rewrite

### PR DESCRIPTION
**Pull request by nickdesaulniers**
mostly stylistic changes other than the meat in src/poll_ctx.cc.

### CODE REVIEW

File: src/poll_ctx.cc [hunk 1]
```diff
@@ -1,48 +1,35 @@
...
-  if (events & UV_READABLE) {
-    context->invoke_callback(events);
-  }
+void PollCtx::on_readable(uv_poll_t* req, int /* status */, int events) {
+  if (!(events & UV_READABLE))
+    return;
+  Nan::HandleScope scope;
+  v8::Local<v8::Value> argv[] = { Nan::New<v8::Number>(events) };
+  PollCtx* ctx = reinterpret_cast<PollCtx*>(req->data);
+  ctx->callback->Call(1, argv, ctx->async_resource);
```
Comment (reqshark) [inline src/poll_ctx.cc]: very nice

File: src/poll_ctx.cc [hunk 2]
```diff
@@ -1,48 +1,35 @@
...
-
-static void NanomsgReadable(uv_poll_t* req, int /* status */, int events) {
-  const PollCtx* const context = static_cast<PollCtx*>(req->data);
-  if (events & UV_READABLE) {
-    context->invoke_callback(events);
-  }
+void PollCtx::on_readable(uv_poll_t* req, int /* status */, int events) {
+  if (!(events & UV_READABLE))
+    return;
+  Nan::HandleScope scope;
```
Comment (reqshark) [inline src/poll_ctx.cc]: i like that, +1 for context avoidance

**Comment by reqshark** [comment]
@nickdesaulniers, this looks good, I like the style stuff. As for the pollctxing, let me poke around at this, give the new code to a long-running process just to make sure memory is nice and flat etc etc

**Comment by nickdesaulniers** [comment]
I honestly wish I had written a memory leak test the last time we looked into it a few years ago. :(

**Pull request merged**